### ISSUE TITLE
Re-register deregistered container instance 

### ISSUE THREAD

**Issue by tj** [aws/amazon-ecs-agent#125]
Since `register-container-instance` is "private", is there another way to re-register a container instance? Primarily for maintenance purposes. We'll probably just end up working around this by making our instances easier to toss away and spin up new ones, but there may still be valid use-cases where you'd just like to momentarily de-register.

The docs are also not clear on the method of which the running containers are shut down when de-registering a container instance. Is this graceful? Are they still properly deregistered from ELBs and SIGTERM'd or are they SIGKILL'd etc

thanks!

**Comment by euank** [comment]
De-registering is supposed to be final.
If you would like to register as a new container instance, you can remove the agent's checkpointed data (at `/var/lib/ecs/data/*` by default) before starting the agent, but all previously managed containers will be forgotten about / 'orphaned' as well.

You're supposed to stop all tasks on a container instance before deregistering it (and the API won't let you succeed unless you have done so or use the force flag), and `stop-task` does send a `SIGTERM`. There's no way via the current APIs to cause a `SIGKILL` to be sent.
If you do use the `--force` flag on `deregister-container-instance`, it will move all containers on it to stopped, regardless of all else, and move it to inactive. I do not believe there is a way to stop 'orphaned' tasks after forcefully deregister a container instance as the associated container agent will not be able to interact with the ECS service any longer.

With regards to being deregistered from ELBs, in either case they should be because the service scheduler will see them as 'stopped'.

Once it's in deregistered, the data (e.g. tasks and container instance) are eligible for being removed after a short grace period I believe.

There is not a way to re-register a container instance once it's deregistered and, given the above, I don't think it makes sense to have that.

If you do have a use-case here, I think it's likely we're really missing something somewhere else. Perhaps there needs to be a way to signal that a container instance should not be placed on temporarily (similar to a draining state).

Apologies for my response being a bit scattered; hopefully it's still coherent.

I'd really like more information on what use-cases you're thinking of and whether you have any particular behavior in mind.
Thanks,
Euan

**Comment by tj** [comment]
Ahh thanks that clears things up, the documentation just mentions the following which seemed a little ambiguous, --force is usually an aggressive thing, but if it just does the stops for you that sounds good to me.

> Force the deregistration of the container instance. You can use the force parameter if you have several tasks running on a container instance and you don't want to run stop-task for each task before deregistering the container instance.

As far as the use-cases go, the primary one was to deregister so that nothing gets scheduled to the instance, giving us a chance to upgrade Docker to 1.7, then re-registering. Not the biggest deal since we can spin up new instances, it's just a little awkward with Terraform.

Feel free to close I was mostly curious about some clarity on the semantics there, very helpful thanks!

**Comment by euank** [comment]
I'm closing this one as the feature-request related to this is more explicitly expressed in #130 and I think your original questions got cleared up. Feel free to reopen with more questions or add in more comments on that new issue as relevant.

Thanks,
Euan

**Comment by mahmah2** [comment]
@euank In our case the EC2 instance size was not enough so I upgraded the instance size but even after upgrade the size was not updated in ECS infrastructure section so I had to deregister the external container instance and now that I want to re-register it again it doesn't appear in the external container instances. Should I wait for some time and re-register it again or I should create another EC2 instance ?

**Comment by sergiy-kozak** [comment]
> @euank In our case the EC2 instance size was not enough so I upgraded the instance size but even after upgrade the size was not updated in ECS infrastructure section so I had to deregister the external container instance and now that I want to re-register it again it doesn't appear in the external container instances. Should I wait for some time and re-register it again or I should create another EC2 instance ?

Hi, did you get by chance any workaround/recipe for the re-registering like you have described?

**Comment by d3dfx** [comment]
For those who come to this page again. This worked for me!

```
systemctl stop ecs
rm -f /var/lib/ecs/*
systemctl start ecs
```
